README.md
CrontabManager
==============

PHP library for GNU/Linux cron jobs management.

Some features:

-   Deal with your cron jobs in PHP
-   Create new Cron jobs
-   Update existing cron jobs
-   Enable / disable cron jobs
-   Manage cron jobs of others users than runtime user using some sudo
    tricks (see below)

Requirements:
-------------

-   PHP 5.3+
-   `crontab` command-line utility (installed by default on common Linux
    distribution)
-   `sudo`, if you want to manage crontab of another user than runtime
    user without running into right issues (see below)

Installation:
-------------

The library can be installed using Composer.

    composer require tiben/crontab-manager ~1.0

Usage:
------

The library is composed of three classes:

-   `CrontabJob` is an entity class that represents a cron job.
-   `CrontabRepository` is used to persist and or retrieve your cron
    jobs.
-   `CrontabAdapter` handles cron jobs persistance in the crontab.

### Instanciate the repository:

In order to work, the CrontabRepository needs an instance of
CrontabAdapter.

``` {.php}
$crontabRepository = new CrontabRepository(new CrontabAdapter());
```

### Create new Job and persist it into the crontab

Suppose you want to create an new job which consists of launching the
command `df >> /tmp/df.log` every day at 23:30. There is two ways this
can be done:

-   In Pure `OO` way :

    ``` {.php}
    $crontabJob = new CrontabJob();
    $crontabJob->minutes = '30';
    $crontabJob->hours = '23';
    $crontabJob->dayOfMonth = '*';
    $crontabJob->months = '*';
    $crontabJob->dayOfWeek = '*';
    $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    $crontabJob->comments = 'Logging disk usage'; // Comments are persisted in the crontab
    ```

-   From raw cron syntax string using a factory method :

    ``` {.php}
    $crontabJob = CrontabJob::createFromCrontabLine('30 23 * * * df >> /tmp/df.log');
    ```

You can now add your new cron job in the crontab repository and persist
all changes to the crontab.

``` {.php}
$crontabRepository->addJob($crontabJob);
$crontabRepository->persist();
```

### Find a specific cron job from the crontab repository and update it

Suppose we want to modify the hour of an already existing cronjob.
Finding existing cron jobs is done using regular expressions. The regex
is applied to the entire crontab line.

``` {.php}
$results = $crontabRepository->findJobByRegex('/Logging\ disk\ usage/');
$crontabJob = $results[0];
$crontabJob->hours = '21';
$crontabRepository->persist();
```

### Remove a cron job from the crontab

You can remove a job like this:

``` {.php}
$results = $crontabRepository->findJobByRegex('/Logging\ disk\ usage/');
$crontabJob = $results[0];
$crontabRepository->removeJob($crontabJob);
$crontabRepository->persist();
```

Note: Since cron jobs are internally matched by reference, they must be
previously obtained from the repository or previously added.

### Work with the crontab of another user than runtime user

This feature allows you to manage the crontab of another (unix) user
than the user who launched the php runtime. This can be useful when the
runtime user is `www-data` — common when PHP is used along with a
webserver — but the owner of the crontab you want to edit is your own
linux user account.

To do this, simply pass the username of the crontab owner as a parameter
of the CrontabAdapter constructor. Suppose you are `www-data` and you
want to edit the crontab of user `bobby`:

``` {.php}
$crontabAdapter = new CrontabAdapter('bobby');
$crontabRepository = new CrontabRepository($crontabAdapter);
```

In this context you will propably run into user rights issues. This can
be handled by editing your sudoers file using 'visudo'.\
If you want to allow user `www-data` to edit the crontab of user
`bobby`, add this line:

    www-data        ALL=(bobby) NOPASSWD: /usr/bin/crontab

which tells sudo not to ask for password when user `www-data` calls
`crontab` as user `bobby` using `sudo`

Now, you can access to the crontab of user `bobby` like this :

``` {.php}
$crontabAdapter = new CrontabAdapter('bobby', true);
$crontabRepository = new CrontabRepository($crontabAdapter);
```

**Note the second parameter `true`** of the CrontabAdapter constructor
call. This boolean tells the CrontabAdapter to use `sudo` internally
when calling `crontab`.

### Enable or disable a cron job

You can enable or disable your cron jobs by setting the `enabled`
attribute of a CronJob object accordingly :

``` {.php}
$crontabJob->enabled = false;
```

This will have the effect to prepend your cron job by a "\#" in your
crontab when persisting it.


CHANGELOG.md
1.2.2 (2018-09-25)
------------------
- Make tests compatible with PHPUnit 6+
- Fix #6 (Thanks @wenqinruan)

1.2.1 (2016-08-27)
------------------
- Fixed findJobByRegex not searching comments

1.2.0 (2014-05-17)
------------------
- Cronjob can now be enabled/disabled

1.1.0 (2014-01-11)
------------------
- Cronjob can now be removed from de crontab

1.0.0 (2014-01-11)
------------------
- Done some refactoring and cs fixes 
- Written README.md documentation
- Preparing source to be published on Github



test/CrontabJobTest.php
<?php
/**
 * CrontabJob class PHPUnit test cases
 * @author FranceProNet
 */
require_once dirname(__FILE__) . '/../../CrontabManager/CrontabJob.php';
	
class CrontabJobTest extends PHPUnit_Framework_TestCase {
	
	protected function setUp() {

	}	
	
	/**
	 * Test parsing crontab and creation of cronjob 
	 * @dataProvider crontabLines
	 */
	public function testCreateACrontabJobFromCrontabLine($crontabLine, $crontabJob) {
		
		$this->assertEquals($crontabJob, CrontabJob::createFromCrontabLine($crontabLine));
		
	}	
	
    public static function crontabLines()  {
    	
    	$crontabLines = array();
    	
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '30';
    	$crontabJob->hours = '23';
    	$crontabJob->dayOfMonth = '*';
    	$crontabJob->months = '*';
    	$crontabJob->dayOfWeek = '*';
    	$crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    	
    	$crontabLines[] = array('30 23 * * * df >> /tmp/df.log', $crontabJob);

    	
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '12';
    	$crontabJob->hours = '10';
    	$crontabJob->dayOfMonth = '2-5';
    	$crontabJob->months = '*';
    	$crontabJob->dayOfWeek = '*';
    	$crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    	
    	$crontabLines[] = array('12 10 2-5 * * df >> /tmp/df.log', $crontabJob);

    	
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '59';
    	$crontabJob->hours = '23';
    	$crontabJob->dayOfMonth = '*/2';
    	$crontabJob->months = '*';
    	$crontabJob->dayOfWeek = '*';
    	$crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    	
    	$crontabLines[] = array('59 23 */2 * * df >> /tmp/df.log', $crontabJob);
    	
    	
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '0';
    	$crontabJob->hours = '0';
    	$crontabJob->dayOfMonth = '25-31';
    	$crontabJob->months = '1,3,5,7,8,10,12';
    	$crontabJob->dayOfWeek = '0';
    	$crontabJob->taskCommandLine = 'my-script.sh';
    	
    	$crontabLines[] = array('0 0 25-31 1,3,5,7,8,10,12 0 my-script.sh', $crontabJob);
    	
    	/* space at start test case */
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '0';
    	$crontabJob->hours = '*';
    	$crontabJob->dayOfMonth = '*/2';
    	$crontabJob->months = '*';
    	$crontabJob->dayOfWeek = '*';
    	$crontabJob->taskCommandLine = 'my-script.sh';
    	$crontabJob->comments = 'commentaire blablabla';
    	
    	$crontabLines[] = array(' 0 * */2 * * my-script.sh #commentaire blablabla', $crontabJob);

    	/* Tab or spaces test case */
    	$crontabJob = new CrontabJob();    	
    	$crontabJob->minutes = '0';
    	$crontabJob->hours = '*';
    	$crontabJob->dayOfMonth = '*/2';
    	$crontabJob->months = '*';
    	$crontabJob->dayOfWeek = '*';
    	$crontabJob->taskCommandLine = 'my-script.sh';
    	$crontabJob->comments = 'commentaire blablabla';
    	
    	$crontabLines[] = array('0	*	 */2	  *		  *		 my-script.sh #commentaire blablabla', $crontabJob);
    	
    	/* Cron shortcut test case */
    	$crontabJob = new CrontabJob();
    	$crontabJob->shortCut = 'daily';
    	$crontabJob->taskCommandLine = 'my-script.sh';
    	
    	$crontabLines[] = array('@daily my-script.sh', $crontabJob);    	
    	
    	return $crontabLines;

    }
    
    public function testFormatCrontabLine() {

    	/* Normal formated crontab line */
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '30';
    	$crontabJob->hours = '23';
    	$crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    	 
    	$this->assertEquals('30 23 * * * df >> /tmp/df.log', $crontabJob->formatCrontabLine());
    	
    	$crontabJob = new CrontabJob();
    	
    	/* Shortcut formated crontab line */
    	$crontab = new CrontabJob();
    	$crontabJob->shortCut = 'reboot';
    	$crontabJob->taskCommandLine = 'echo "youpi"';
    	
    	$this->assertEquals('@reboot echo "youpi"', $crontabJob->formatCrontabLine());
    	
    	/* crontab line with comments */
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '30';
    	$crontabJob->hours = '23';
    	$crontabJob->taskCommandLine = 'df >> /tmp/df.log';
    	$crontabJob->comments = 'This job is commented';
    	 
    	$this->assertEquals('30 23 * * * df >> /tmp/df.log #This job is commented', $crontabJob->formatCrontabLine());    	
    	
    }
    
    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
	public function testCrontabLineNotWellFormatedNotEnoughStars() {
		CrontabJob::createFromCrontabLine('* * 15 * youpi');		
	}    

	/**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
	public function testCrontabLineNotWellFormatedShortcutWithoutArobase() {
		CrontabJob::createFromCrontabLine('daily youpi');		
	}    

	/**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
	public function testCrontabLineNotWellFormatedUnknownShortcut() {
		CrontabJob::createFromCrontabLine('@maintenant youpi');		
	}    
    
    
    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage CrontabJob contain's no task command line
     */
    public function testNoTaskSetException()  {
    	
    	$crontabJob = new CrontabJob();
    	$crontabJob->minutes = '30';
    	$crontabJob->hours = '23';

		$crontabJob->formatCrontabLine();    	
    	
    }
	
}		

?>

test/CrontabRepositoryTest.php
<?php
/**
 * CrontabRepository class PHPUnit test cases
 * @author FranceProNet
 */
require_once dirname(__FILE__) . '/../../CrontabManager/CrontabRepository.php';
	
class CrontabRepositoryTest extends PHPUnit_Framework_TestCase {
	
	protected $fixturesPath;
	
	protected function setUp() {
		$this->fixturesPath = dirname(__FILE__) . '/Fixtures/';
	}	
	
	/**
	 * Test if the crontab file is parsed.
	 */
	public function testParseCrontab() {

		/* Create fake crontabAdapter */
		$fakeCrontabAdapter = $this->getMock('CrontabAdapter');
		$fakeCrontabAdapter
			->expects($this->any())
			->method('readCrontab')
			->will($this->returnValue(file_get_contents($this->fixturesPath . 'testing_read_crontab.txt')))
		;		
		
		/* Create expected crontabJobs */
		
		$expectedCrontabJob1 = new CrontabJob();
		$expectedCrontabJob1->minutes = '30';
		$expectedCrontabJob1->hours = '23';
		$expectedCrontabJob1->dayOfMonth = '*';
		$expectedCrontabJob1->months = '*';
		$expectedCrontabJob1->dayOfWeek = '*';
		$expectedCrontabJob1->taskCommandLine = 'df >> /tmp/df.log';		
		$expectedCrontabJob1->comments = 'first crontabJob';
		
		$expectedCrontabJob2 = new CrontabJob();
		$expectedCrontabJob2->minutes = '0';
		$expectedCrontabJob2->hours = '0';
		$expectedCrontabJob2->dayOfMonth = '28-31';
		$expectedCrontabJob2->months = '*';
		$expectedCrontabJob2->dayOfWeek = '*';
		$expectedCrontabJob2->taskCommandLine = '[ `/bin/date +\%d` -gt `/bin/date +\%d -d "1 day"` ] && df >> /tmp/df.log';		
		$expectedCrontabJob2->comments = 'second crontabJob';

		$expectedCrontabJob3 = new CrontabJob();
		$expectedCrontabJob3->shortCut = 'hourly';
		$expectedCrontabJob3->taskCommandLine = 'df > /tmp/df_`date +\%d_\%m_\%Y_\%H_\%M`.log';		
		$expectedCrontabJob3->comments = 'third crontabJob';

		$expectedCrontabJobs = array($expectedCrontabJob1, $expectedCrontabJob2, $expectedCrontabJob3);
		
		$crontabRepository = new CrontabRepository($fakeCrontabAdapter);
		$crontabJobs = $crontabRepository->getJobs();
			
		$this->assertEquals($expectedCrontabJobs, $crontabJobs);
		
	}
	
	/**
	 * Test if the headers Comments of the crontab file are read
	 */	
	public function testReadHeaderComments() {
		
		/* Create fake crontabAdapter */
		$fakeCrontabAdapter = $this->getMock('CrontabAdapter');
		$fakeCrontabAdapter
			->expects($this->any())
			->method('readCrontab')
			->will($this->returnValue(file_get_contents($this->fixturesPath . 'simple_crontab.txt')))
		;		
		
		$crontabRepository = new CrontabRepository($fakeCrontabAdapter);
		
		$this->assertEquals(file_get_contents($this->fixturesPath . 'crontab_headers.txt'), $crontabRepository->headerComments);
		
	}
	
	/**
	 * Test finding a job by a regular expression 
	 */	
	public function testFindJobByRegex() {
		
		/* Create fake crontabAdapter */
		$fakeCrontabAdapter = $this->getMock('CrontabAdapter');
		$fakeCrontabAdapter
			->expects($this->any())
			->method('readCrontab')
			->will($this->returnValue(file_get_contents($this->fixturesPath . 'simple_crontab.txt')))
		;
		
		$crontabJob = new CrontabJob();
		$crontabJob->minutes = '30';
		$crontabJob->hours = '23';
		$crontabJob->dayOfMonth = '*';
		$crontabJob->months = '*';
		$crontabJob->dayOfWeek = '*';
		$crontabJob->taskCommandLine = 'launch -param mycommand';		
		
		$expectedCrontabJobs = array($crontabJob);
		
		$crontabRepository = new CrontabRepository($fakeCrontabAdapter);		
		$crontabJobs = $crontabRepository->findJobByRegex('/launch -param mycommand/');
			
		$this->assertEquals($expectedCrontabJobs, $crontabJobs);
				
	}
	
	/**
	 * This test will modify an existing job and append a new job to the crontab.
	 */
	public function testPersist() {

		/* Create fake crontabAdapter */
		$fakeCrontabAdapter = $this->getMock('CrontabAdapter');
		
		$fakeCrontabAdapter
			->expects($this->any())
			->method('readCrontab')
			->will($this->returnValue(file_get_contents($this->fixturesPath . 'simple_crontab.txt')))			
		;
		
		$fakeCrontabAdapter			
			->expects($this->once())
			->method('writeCrontab')
			->with($this->equalTo(file_get_contents($this->fixturesPath . 'testing_persisted_crontab.txt')))
		;
		
		$crontabRepository = new CrontabRepository($fakeCrontabAdapter);
		
		/* Modify the existing job */
		$crontabJobs = $crontabRepository->findJobByRegex('/launch -param mycommand/');
		$crontabJobs[0]->minutes = '01';
		$crontabJobs[0]->hours = '05';
		
		/* Append new job */
		$newCrontabJob = new CrontabJob();
		$newCrontabJob->minutes = '30';
		$newCrontabJob->hours = '23';
		$newCrontabJob->taskCommandLine = 'df >> /tmp/df.log';
		$newCrontabJob->comments = 'new crontab job';
		
		$crontabRepository->addJob($newCrontabJob);
		
		$crontabRepository->persist();
		
	}
	
	/**
	 * Test if pass a wrong regular expression when searching by regex throw an invalid regex exception
	 * @expectedException InvalidArgumentException
	 * @expectedExceptionMessage Not a valid Regex : preg_match(): No ending delimiter '/' found
	 */
	public function testExceptionInvalidRegexOnFindJobByRegex() {

		$fakeCrontabAdapter = $this->getMock('CrontabAdapter');
		
		$fakeCrontabAdapter
			->expects($this->any())
			->method('readCrontab')
			->will($this->returnValue(file_get_contents($this->fixturesPath . 'simple_crontab.txt')))
		;		
		$crontabRepository = new CrontabRepository($fakeCrontabAdapter);
		
		$crontabRepository->findJobByRegex('/$');
		
	}
	
}	

tests/CrontabManager/Tests/CrontabJobTest.php
<?php

/*
 * Copyright 2013 Benjamin Legendre
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace TiBeN\CrontabManager\Tests;

use TiBeN\CrontabManager\CrontabJob;

/**
 * CrontabJob class PHPUnit test cases
 *
 * @author TiBeN
 */
    
class CrontabJobTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test parsing crontab and creation of cronjob 
     * 
     * @dataProvider crontabLines
     */
    public function testCreateACrontabJobFromCrontabLine($crontabLine, $crontabJob)
    {
        $this->assertEquals($crontabJob, CrontabJob::createFromCrontabLine($crontabLine));
    }
    
    public static function crontabLines()
    {
        $crontabLines = array();
        
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->dayOfMonth = '*';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabLines[] = array('30 23 * * * df >> /tmp/df.log', $crontabJob);
        
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '12';
        $crontabJob->hours = '10';
        $crontabJob->dayOfMonth = '2-5';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabLines[] = array('12 10 2-5 * * df >> /tmp/df.log', $crontabJob);
        
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '59';
        $crontabJob->hours = '23';
        $crontabJob->dayOfMonth = '*/2';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabLines[] = array('59 23 */2 * * df >> /tmp/df.log', $crontabJob);
        
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '0';
        $crontabJob->hours = '0';
        $crontabJob->dayOfMonth = '25-31';
        $crontabJob->months = '1,3,5,7,8,10,12';
        $crontabJob->dayOfWeek = '0';
        $crontabJob->taskCommandLine = 'my-script.sh';
        $crontabLines[] = array('0 0 25-31 1,3,5,7,8,10,12 0 my-script.sh', $crontabJob);
        
        /* space at start test case */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '0';
        $crontabJob->hours = '*';
        $crontabJob->dayOfMonth = '*/2';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'my-script.sh';
        $crontabJob->comments = 'commentaire blablabla';
        $crontabLines[] = array(' 0 * */2 * * my-script.sh #commentaire blablabla', $crontabJob);

        /* Tab or spaces test case */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->minutes = '0';
        $crontabJob->hours = '*';
        $crontabJob->dayOfMonth = '*/2';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'my-script.sh';
        $crontabJob->comments = 'commentaire blablabla';
        $crontabLines[] = array('0  *    */2      *       *      my-script.sh #commentaire blablabla', $crontabJob);
        
        /* Cron shortcut test case */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = true;
        $crontabJob->shortCut = 'daily';
        $crontabJob->taskCommandLine = 'my-script.sh';
        $crontabLines[] = array('@daily my-script.sh', $crontabJob);
        
        /* Disabled cron test case - without spaces */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = false;
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->dayOfMonth = '*';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabLines[] = array('#30 23 * * * df >> /tmp/df.log', $crontabJob);

        /* Disabled cron test case - with spaces */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = false;
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->dayOfMonth = '*';
        $crontabJob->months = '*';
        $crontabJob->dayOfWeek = '*';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabLines[] = array('#   30 23 * * * df >> /tmp/df.log', $crontabJob);

        return $crontabLines;
    }
    
    public function testFormatCrontabLine()
    {
        /* Normal formated crontab line */
        $crontabJob = new CrontabJob();
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $this->assertEquals('30 23 * * * df >> /tmp/df.log', $crontabJob->formatCrontabLine());
        $crontabJob = new CrontabJob();
        
        /* Shortcut formated crontab line */
        $crontab = new CrontabJob();
        $crontabJob->shortCut = 'reboot';
        $crontabJob->taskCommandLine = 'echo "youpi"';
        $this->assertEquals('@reboot echo "youpi"', $crontabJob->formatCrontabLine());
        
        /* crontab line with comments */
        $crontabJob = new CrontabJob();
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabJob->comments = 'This job is commented';
        $this->assertEquals(
            '30 23 * * * df >> /tmp/df.log #This job is commented', 
            $crontabJob->formatCrontabLine()
        );

        /* disabled crontab line */
        $crontabJob = new CrontabJob();
        $crontabJob->enabled = false;
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->taskCommandLine = 'df >> /tmp/df.log';
        $crontabJob->comments = 'This job is commented';
        $this->assertEquals(
            '#30 23 * * * df >> /tmp/df.log #This job is commented', 
            $crontabJob->formatCrontabLine()
        );

    }
    
    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
    public function testCrontabLineNotWellFormatedNotEnoughStars()
    {
        CrontabJob::createFromCrontabLine('* * 15 * youpi');
    }

    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
    public function testCrontabLineNotWellFormatedShortcutWithoutArobase()
    {
        CrontabJob::createFromCrontabLine('daily youpi');
    }

    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Crontab line not well formated then can't be parsed
     */
    public function testCrontabLineNotWellFormatedUnknownShortcut()
    {
        CrontabJob::createFromCrontabLine('@maintenant youpi');
    }
    
    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage CrontabJob contain's no task command line
     */
    public function testNoTaskSetException()
    {
        $crontabJob = new CrontabJob();
        $crontabJob->minutes = '30';
        $crontabJob->hours = '23';
        $crontabJob->formatCrontabLine();
    }
}


src/CrontabManager/CrontabAdapter.php
<?php

/*
 * Copyright 2013 Benjamin Legendre
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace TiBeN\CrontabManager;

/**
 * Crontab adapter. 
 * Retrieve and write cron jobs data using the "crontab" command line.
 * @author Benjamin Legendre
 */
class CrontabAdapter
{
    private $userName;
    
    private $useSudo;
    
    /**
     * Instantiate a crontabAdapter
     * 
     * @param String $userName Optional Tell the crontab 
     * user where the adapter will try to read (by default user = runtime user)
     * @param Boolean $useSudo Tell Adapter use sudo command to connect 
     * to crontab of another user than the runtime user.
     *  
     * About sudo :
     * If you want to work with the crontab of another user 
     * of the runtime user, you can allowing the runtime user to edit 
     * another user crontab by adding this kind of line in your sudoers using 'visudo'
     * For example : user www-data want to edit waylon crontab:
     * 
     * www-data        ALL=(waylon) NOPASSWD: /usr/bin/crontab 
     * 
     * Will tell sudo user www-data can execute /usr/bin/crontab as 
     * user waylon without typing any password (required in our case) 
     */
    public function __construct($userName = null, $useSudo = false)
    {
        if ($userName) {
            $this->userName = $userName;
        }
        $this->useSudo = $useSudo;
    }
    
    /**
     * Read the crontab and return
     * raw data
     * @return String $output the crontab raw data
     */
    public function readCrontab()
    {
        $crontabCommandLine = (isset($this->userName) && $this->useSudo)
            ? sprintf('sudo -n -u %s crontab -l', $this->userName)
            : ($this->userName ? sprintf('crontab -u %s -l', $this->userName) : 'crontab -l')
        ;
            
        exec($crontabCommandLine . ' 2>&1', $output, $exitCode);
                
        /* exec error handling */
        if ($exitCode !== 0) {
            
            /* Special case : the crontab is empty throw bad exit code but access is ok */
            if (!preg_match('/^no crontab for .+$/', $output[0])) {
                throw new \DomainException('Error when trying to read crontab : ' . implode(' ', $output));
            } else {
                $output = '';
            }
        } else {
            $output = implode("\n", $output);
        }
        
        return $output;
        
    }
    
    /**
     * Write the raw crontab data to the crontab.
     * @param String $crontabRawData
     */
    public function writeCrontab($crontabRawData)
    {
        $crontabRawData = escapeshellarg($crontabRawData);
        
        $crontabCommandLine = (isset($this->userName) && $this->useSudo)
            ? sprintf('echo %s | sudo -n -u %s crontab -', $crontabRawData, $this->userName)
            : ($this->userName
                ? sprintf('echo %s | crontab -u %s -', $crontabRawData, $this->userName)
                : sprintf('echo %s | crontab -', $crontabRawData)
            )
        ;
        
        exec($crontabCommandLine . ' 2>&1', $output, $exitCode);
        
        /* exec error handling */
        if ($exitCode !== 0) { 
            throw new \DomainException('Error when trying to write crontab : ' . implode(' ', $output));
        }
    }
}


src/CrontabManager/CrontabJob.php
<?php

/*
 * Copyright 2013 Benjamin Legendre
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace TiBeN\CrontabManager;

/**
 * CrontabJob
 * Represent a Job of the crontab.
 * 
 * @author Benjamin Legendre
 */
class CrontabJob
{
    /**
     * Tell whether the cron job is enabled or not
     * This will add or not a # at the beginning of the cron line
     *
     * @var boolean
     */
    public $enabled = true;

    /**
     * Min (0 - 59)
     *
     * @var String/int
     *
     */
    public $minutes;
    
    /**
     * Hour (0 - 23)
     *
     * @var String/int
     */
    public $hours;
    
    /**
     * Day of month (1 - 31)
     *
     * @var String/int
     */
    public $dayOfMonth;
    
    /**
     * Month (1 - 12)
     *
     * @var String/int
     */
    public $months;
    
    /**
     * Day of week (0 - 6) (0 or 6 are Sunday to Saturday, or use names)
     *
     * @var String/int
     */
    public $dayOfWeek;

    /**
     * The task command line to be executed 
     *
     * @var String
     */
    public $taskCommandLine;
    
    /**
     * Optional comment that will be placed at the end of the crontab line 
     * and preceded by a #
     *
     * @var String
     */
    public $comments;
    
    /**
     * Predefined scheduling definition
     * Shorcut définition that replace standard définition (preceded by @)
     * possibles values : yearly, monthly, weekly, daily, hourly, reboot
     * When a shortcut is defined, it overwrite stantard définition
     *
     * @var String
     */
    public $shortCut;
    
    /**
     * Factory method to create a CrontabJob from a crontab line.
     *
     * @param String $crontabLine
     * @throws InvalidArgumentException
     * @return CrontabJob
     */
    public static function createFromCrontabLine($crontabLine)
    {
        // Check crontab line format validity
        $crontabLineRegex = '/^[\s\t]*(#)?[\s\t]*(([*0-9,-\/]+)[\s\t]+([*0-9,-\/]+)'
            . '[\s\t]+([*0-9,-\/]+)[\s\t]+([*a-z0-9,-\/]+)[\s\t]+([*a-z0-9,-\/]+)|'
            . '(@(reboot|yearly|annually|monthly|weekly|daily|midnight|hourly)))'
            . '[\s\t]+([^#]+)([\s\t]+#(.+))?$/'
        ;

        if (!preg_match($crontabLineRegex, $crontabLine, $matches)) {
            throw new \InvalidArgumentException(
                'Crontab line not well formated then can\'t be parsed'
            );
        }

        // Create the job from parsed crontab line values
        $crontabJob = new self();
      
        if (!empty($matches[1])) {
            $crontabJob->enabled = false;
        }

        if (!empty($matches)) {
            $crontabJob->minutes = $matches[3];
            $crontabJob->hours = $matches[4];
            $crontabJob->dayOfMonth = $matches[5];
            $crontabJob->months = $matches[6];
            $crontabJob->dayOfWeek = $matches[7];
        }
        
        if (!empty($matches[8])) {
            $crontabJob->shortCut = $matches[9];
        }
        
        $crontabJob->taskCommandLine = $matches[10];
        if (!empty($matches[12])) {
            $crontabJob->comments = $matches[12];
        }
        
        return $crontabJob;
    }
    
    /**
     * Format the CrontabJob to a crontab line 
     *
     * @throws InvalidArgumentException
     * @return String
     */
    public function formatCrontabLine()
    {
        
        // Check if job has a task command line
        if (!isset($this->taskCommandLine) || empty($this->taskCommandLine)) {
            throw new \InvalidArgumentException(
                'CrontabJob contain\'s no task command line'
            );
        }
        
        $taskPlanningNotation = (isset($this->shortCut) && !empty($this->shortCut))
            ? sprintf('@%s', $this->shortCut)
            : sprintf(
                '%s %s %s %s %s',
                (isset($this->minutes) ? $this->minutes : '*'),
                (isset($this->hours) ? $this->hours : '*'),
                (isset($this->dayOfMonth) ? $this->dayOfMonth : '*'),
                (isset($this->months) ? $this->months : '*'),
                (isset($this->dayOfWeek) ? $this->dayOfWeek : '*')
            )
        ;
        
        return sprintf(
            '%s%s %s%s',
            ($this->enabled ? '' : '#'),
            $taskPlanningNotation,
            $this->taskCommandLine,
            (isset($this->comments) ? (' #' . $this->comments) : '')
        );
    }
}
